File: 프로그래머스/3/43163. 단어 변환/단어 변환.js
function solution(begin, target, words) {
    const n = words.length;
    if (!words.includes(target)) return 0;
    
    function diff1(a, b) {
        let diff = 0;
        for (let i = 0; i < a.length; i++) {
            if (a[i] !== b[i]) diff++;
        }
        return diff === 1;
    }
    
    let q = [[begin, 0]];
    let head = 0;
    const visited = Array(n).fill(false);
    
    while(head < q.length) {
        const [cur, dist] = q[head++];
        if (cur === target) return dist;
        for (let i = 0; i < n; i++) {
            if (!visited[i] && diff1(cur, words[i])) {
                visited[i] = true;
                q.push([words[i], dist+1]);
            }
        }
    }
    return 0;
}

File: 프로그래머스/3/43163. 단어 변환/단어 변환.py
from collections import deque

def can_convert(w1, w2):
    diff = 0
    
    for c1, c2 in zip(w1, w2):
        if c1 != c2:
            diff += 1
            if diff > 1:
                return False
    
    return diff == 1

def solution(begin, target, words):
    if target not in words:
        return 0
    
    q = deque([(begin, 0)])
    visited = [False] * len(words)
    
    while q:
        cur, cnt = q.popleft()
        
        if cur == target:
            return cnt
        
        for i, word in enumerate(words):
            if not visited[i] and can_convert(cur, word):
                visited[i] = True
                q.append((word, cnt + 1))
    
    return 0